### ISSUE TITLE
Does pubsub_client have async impl ?

### ISSUE THREAD

**Issue by Knight-X** [solana-labs/solana#22006]
#### Problem

Hey Guys, I am using solana pubsub client as part of my program. I found that WS is using tungstenite-rs but without having async impl (tokio-tungstenite). The performance of the pubsub client is blocked by the receiving loop. Morever, the websocket spec doesn't have the ping/pong functionality to have heartbeat check. 

#### Proposed Solution
1. Should I implement the async version of the pubsub client and making the pr in the future or do we have another plan ? 
2. Do we have alternative method to make the heartbeat check ? (such as subscribe to slotNotification)

discussion welcome

**Comment by mvines** [comment]
1 sounds great, I don't know of anybody working on another plan!

**Comment by fanatid** [comment]
Also interested in async clients. Any suggestions on how this should be added? In addition to module `pubsub_client` we should add a new module `pubsub_client_async` with suffixes `Async`/`_async`? Should we put this behind the feature?

**Comment by mvines** [comment]
Carving out an `solana-client::async` module seems decent, where we might have:
* `solana-client::async::rpc_client::RpcClient`
* `solana-client::async::pubsub_client::PubsubClient`

**Comment by fanatid** [comment]
@mvines should I change my PR?

**Comment by mvines** [comment]
I do feel `solana-client::async::{pubsub_client::PubsubClient, rpc_client:RpcClient}` is marginally better than `solana-client::{pubsub_client_async::PubsubClient, rpc_client_async::RpcClient}`  but I'm also very open to other opinions/ideas.

**Comment by fanatid** [comment]
#20683 has a suggestion about the thin layer on top of async code (spinning executor on async code). I think we can do the same for `pubsub`, sounds good?
